chia|tests: Introduce `PlotManager` + some plot loading improvements and fixes

xdustinface opened the pull request:
This PR is fixing and improving harvester plot loading by:

- Wrapping all plot loading related functions/variables in a class called `PlotManager`, 1219f2d3d172869888a006685049f865b5cf910a
- Move plot loading into a separate thread 4c128afe226e5939a604b46426408ccf64f60c34 and load plots in batches 2176a0eca6ac78271d523d7ca9de418cf94e2ac4. This leads to a more responsive harvester during startup which should now also process signage points in time while loading plots.
- Expand plot loading related test coverage (should probably be moved into separate file at some point where it also tests all of `PlotManager`) c3899ba17a5cc1431fbd51aad6f4f6b417589399
- The other commits are fixes or smaller improvements, see individual commit messages.

#### New parameter
There is a new parameter object in the config called `harvester.plots_refresh_parameter` which introduces two new parameter and contains what was previously known as `plot_loading_frequency_seconds`.  
```
plots_refresh_parameter:
    interval_seconds: 120 # The interval in seconds to refresh the plot file manager
    batch_size: 30 # How many plot files the harvester processes before it waits batch_sleep_milliseconds
    batch_sleep_milliseconds: 10 # Milliseconds the harvester sleeps between batch processing
```

xdustinface committed:
harvester|plotting|tests: Introduce `PlotManager` class

This moves all plots related and plot-directory related stuff from the
harvester into the class `PlotManager`, adjusts all related code
accordingly and adds some extra wrappers there.

xdustinface committed:
harvester|plotting|tests: Return how many new plots were loaded

xdustinface committed:
harvester|plotting: Thread locks for `PlotManager.plots`

xdustinface committed:
chia|tests: Load plots in batches

xdustinface committed:
chia|tests: Move plot refreshing into a separate thread

xdustinface committed:
plotting: Fix `failed_to_open_filenames` re-try interval

With `< 1200` it just tries it on the next refresh.

xdustinface committed:
plotting: Fix and improve duplicates handling

xdustinface committed:
plotting: Properly handle removed plots

And fix tests accordingly. It seems like this fix https://github.com/Chia-Network/chia-blockchain/pull/3350 wasn't really a fix, rather adjusting to allow for a bug?

xdustinface committed:
plotting|harvester|tests: Introduce `PlotRefreshResult`

xdustinface committed:
tests: Expand `test_farmer_harvester_rpc.py`

almogdepaz commented on chia/harvester/harvester.py:
```diff
@@ -3,65 +3,58 @@
+        refresh_parameter: PlotsRefreshParameter = PlotsRefreshParameter()
+        if "plot_loading_frequency_seconds" in config:
+            self.log.warning(
+                "plot_loading_frequency_seconds is deprecated but found in config. Replace it with the "
```
i would consider removing this, your setting default params anyway and i think this might scare some new farmers, maybe change from warning to debug

xdustinface replied:
Agree its probably better to not print such a warning, i pushed four more commits and added a migration step to the `load_config` in 5947039889bf16026d6fea9ed0b6af93209c98d3.

almogdepaz replied:
cool but i think we should raise the config migration in the meeting, its useful but could be bad if we mess up users setup, what we did until now is set a hardcoded default for new params

xdustinface replied:
> cool but i think we should raise the config migration in the meeting

Sure we can talk about it later in the meeting :+1:

> but could be bad if we mess up users setup

How do you think it could be messed?

almogdepaz replied:
i suggested the exact same thing few months back and we had the same discussion

almogdepaz replied:
we can mess up the users setup if we corrupt the config file/save a bad value in there

xdustinface replied:
I dropped that migration stuff

almogdepaz commented on chia/plotting/plot_tools.py:
```diff
@@ -106,188 +106,376 @@ def stream_plot_info_ph(
+            while self.needs_refresh() and self._refreshing_enabled:
+                batch_result: PlotRefreshResult = self.refresh_batch()
+                total_result += batch_result
+                if self._refresh_callback is not None:
```
in what case would _refresh_callback be None ?

xdustinface replied:
Whenever `PlotManager` constructor gets created without `refresh_parameter` set like here https://github.com/Chia-Network/chia-blockchain/blob/5947039889bf16026d6fea9ed0b6af93209c98d3/chia/plotting/create_plots.py#L163 or here https://github.com/Chia-Network/chia-blockchain/blob/5947039889bf16026d6fea9ed0b6af93209c98d3/chia/cmds/plots.py#L27

Not that the refreshing gets started in those particular cases but there _might_ be others?

xdustinface replied:
Refactored, see cabfd5d and 88038ff

xdustinface committed:
chia|tests: Move some stuff from `plot_tools.py` into new file `util.py`

xdustinface committed:
refactor: Rename `plot_tools.py` to `manager.py`

xdustinface committed:
chia|tests: Use pure dataclass for `PlotsRefreshParameter`

With `uint16` as type saving to config doesn't work, this is a
preparation for the next commit.

xdustinface committed:
harvester: Adjust deprecation message, use `info` instead of `warning`

xdustinface committed:
plotting: Fix typo

xdustinface committed:
refactor: Rename `filename` to `file_path`

Fits better and does avoid shadowing with filename from outer scope.

xdustinface committed:
plotting: Make `refresh_callback` mandatory

xdustinface committed:
chia|tests: Move some methods from `plotting.manager` to `plotting.util`

almogdepaz commented on chia/plotting/manager.py:
```diff
@@ -0,0 +1,333 @@
+    _lock: threading.Lock
+    _refresh_thread: Optional[threading.Thread]
+    _refreshing_enabled: bool
+    _refresh_callback: Optional[Callable]
```
i think now we can remove the optional from here as well, and the None check in line 127

xdustinface replied:
Dropping it leads to `error: Cannot assign to a method` by `mypy` for the constructor and `set_refresh_callback` so we need to keep it unless you have a way to make `mypy` happy without it.

almogdepaz replied:
in that case i would change the if in line 127 to an assert before pushing

xdustinface replied:
Cool, changed it :+1:

almogdepaz reviewed (approved):
nice!
